folioclient/FolioClient.py

```diff
@@ -161,18 +168,69 @@ def authority_source_files(self):
             )
         )
 
+    @property
+    def okapi_headers(self):
+        """Property that returns okapi headers with the current valid Okapi token"""
+        headers = {
+            "x-okapi-token": self.okapi_token,
+        }
+        headers.update(self.base_headers)
+        return headers
+
+    @property
+    def okapi_token(self):
+        """Property that attempts to return a valid Okapi token, refreshing if needed"""
+        if datetime.now(tz.utc) > (self.okapi_token_expires - timedelta(seconds=self.okapi_token_duration.total_seconds() * .2)):
+            self.refresh()
+        return self._okapi_token
+
+    @okapi_token.setter
+    def okapi_token(self, value):
+        self._okapi_token = value
+
+    @okapi_token.deleter
+    def okapi_token(self):
+        del self._okapi_token
+
+    def refresh(self):
+        """Refreshes okapi_token using the refresh token, when possible"""
+        logging.info("Refreshing okapi token...")
+        url = f"{self.okapi_url}/authn/refresh"
+        req = httpx.post(url, headers=self.base_headers, cookies=self.cookies)
+        if req.status_code == 201:
+            response_body = req.json()
+            self.okapi_token = req.cookies.get("folioAccessToken")
+            self.refresh_token = req.cookies.get("folioRefreshToken")
+            self.cookies = req.cookies
+            self.okapi_token_expires = date_parse(response_body.get("accessTokenExpiration"))
+            self.okapi_token_duration = self.okapi_token_expires - datetime.now(tz.utc)
+            self.refresh_token_expires = date_parse(response_body.get("refreshTokenExpiration"))
+            self.refresh_token_duration = self.refresh_token_expires - datetime.now(tz.utc)
+        else:
+            self.login()
+
     def login(self):
         """Logs into FOLIO in order to get the okapi token"""
         payload = {"username": self.username, "password": self.password}
-        headers = {
-            "x-okapi-tenant": self.tenant_id,
-            "content-type": "application/json",
-        }
-        url = f"{self.okapi_url}/authn/login"
-        req = httpx.post(url, json=payload, headers=headers, timeout=None, verify=self.ssl_verify)
+        # Transitional implementation to support Poppy and pre-Poppy authentication
+        url = f"{self.okapi_url}/authn/login-with-expiry"
+        try:
+            # Poppy and later
+            req = httpx.post(url, json=payload, headers=self.base_headers, timeout=None, verify=self.ssl_verify)
+            req.raise_for_status()
```

Comment:
This would cause a retry of the pre-Poppy authentication regardless of the reason for the failure. Assuming there's a way to distinguish, it seems there should be a distinction between the new endpoint not begin active vs a legit authorization failure. 
Are there cases where it's meaningful to understand why a `login()` or refresh failed, where the failover would obscure the needed correction?

Reply:
I could probably distinguish between a 404 and any other error, and just let the others propogate.